listenbrainz/mbid_mapping/reports/tracks_of_the_year.py

```diff
@@ -0,0 +1,122 @@
+from datetime import datetime, timezone
+
+import click
+import psycopg2
+import psycopg2.extras
+from psycopg2.errors import OperationalError
+
+from mapping.utils import insert_rows, log
+import config
+
+BATCH_SIZE = 5000
+
+
+def create_table(mb_conn):
+    """ Create the tracks of the year table in the mapping schema of a docker-musicbrinz
+        instance. """
+
+    try:
+        with mb_conn.cursor() as curs:
+            curs.execute("DROP TABLE IF EXISTS mapping.tracks_of_the_year")
+            curs.execute("""CREATE TABLE mapping.tracks_of_the_year
+                                       ( user_name          TEXT NOT NULL
+                                       , recording_mbid     UUID NOT NULL
+                                       , recording_name     TEXT NOT NULL
+                                       , artist_credit_name TEXT NOT NULL
+                                       , artist_mbids       UUID[] NOT NULL
+                                       , listen_count       INTEGER NOT NULL
+                                       )""")
+            mb_conn.commit()
+    except (psycopg2.errors.OperationalError, psycopg2.errors.UndefinedTable) as err:
+        log("mbid mapping: failed to create tracks of the year table", err)
+        mb_conn.rollback()
+        raise
+
+
+def create_indexes(mb_conn):
+    """ Create the user_name index on the tracks of the year table. """
+
+    try:
+        with mb_conn.cursor() as curs:
+            curs.execute("""CREATE INDEX tracks_of_the_year_ndx_user_name
+                                      ON mapping.tracks_of_the_year (user_name)""")
+            mb_conn.commit()
+    except (psycopg2.errors.OperationalError, psycopg2.errors.UndefinedTable) as err:
+        log("mbid mapping: failed to create tracks of the year indexes", err)
+        mb_conn.rollback()
+        raise
+
+
+def fetch_user_list(mb_conn, year):
+    """ Fetch the distinct list of users from tracks listened table """
+
+    with mb_conn.cursor(cursor_factory=psycopg2.extras.DictCursor) as mb_curs:
+        query = """SELECT DISTINCT user_name
+                     FROM mapping.tracks_of_the_year"""
+        mb_curs.execute(query)
+        rows = mb_curs.fetchall()
+
+    return [r[0] for r in rows]
+
+
+def chunks(lst, n):
+    """ Helper function to break a list into chunks """
+    for i in range(0, len(lst), n):
+        yield lst[i:i + n]
+
+
+def fetch_tracks_listened_to(lb_conn, mb_conn, ts):
```

Comment:
missing 1 arg and different name from one used inside function. should probably be `lb_conn, mb_conn, start_ts, end_ts`.

Reply:
Shit, I guess I didn't actually check both scripts. Sigh.